gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ProcessWorkUnitsWorkflowImpl.java

```diff
@@ -72,10 +73,33 @@ private CommitStats performWork(WUProcessingSpec workSpec) {
     searchAttributes = TemporalWorkFlowUtils.generateGaasSearchAttributes(jobState.getProperties());
 
     NestingExecWorkflow<WorkUnitClaimCheck> processingWorkflow = createProcessingWorkflow(workSpec, searchAttributes);
-    int workunitsProcessed =
-        processingWorkflow.performWorkload(WorkflowAddr.ROOT, workload, 0, workSpec.getTuning().getMaxBranchesPerTree(),
-            workSpec.getTuning().getMaxSubTreesPerTree(), Optional.empty());
-    if (workunitsProcessed > 0) {
+
+    Optional<Integer> workunitsProcessed = Optional.empty();
+    try {
+     workunitsProcessed = Optional.of(processingWorkflow.performWorkload(WorkflowAddr.ROOT, workload, 0,
+          workSpec.getTuning().getMaxBranchesPerTree(), workSpec.getTuning().getMaxSubTreesPerTree(), Optional.empty()));
+    } catch (Exception e) {
+      log.error("Exception occurred in performing workload,proceeding with commit step", e);
+      // We want to mark the GaaS flow as failure, in case performWorkFlow fails, but we still want to go ahead with commiting the workunits which were processed before failure
+      sendFailureEventToGaaS(workSpec);
+      return proceedWithCommitStepAndReturnCommitStats(workSpec, searchAttributes, workunitsProcessed);
+    }
+    return proceedWithCommitStepAndReturnCommitStats(workSpec, searchAttributes, workunitsProcessed);
+  }
+
+  private void sendFailureEventToGaaS(WUProcessingSpec workSpec) {
```

Comment:
I would advise against mentioning gaas in the name (maybe instead `sendJobFailedEvent`)...

... but the bigger issue is that we don't want to send a JobFailed GTE here and then successfully complete this sub-workflow.  if we did, the parent (`ExecuteGobblinWorkflowImpl`) [will send a JobSuccess GTE](https://github.com/apache/gobblin/blob/4e38b2fd25eb4ca84227f654a0f949c8f598f75e/gobblin-temporal/src/main/java/org/apache/gobblin/temporal/ddm/workflow/impl/ExecuteGobblinWorkflowImpl.java#L134) and the two would be in conflict.  instead we should fail this sub-workflow by throwing an exception

Reply:
addressed